### ISSUE TITLE
Error when using with React

### ISSUE THREAD

**Issue by phamquyhai** [xenova/transformers.js#41]
I tested with React but got error
Code:
```
import { pipeline } from "@xenova/transformers";
....
const classifier = await pipeline("translation");
```

Error:
```
tokenizers.js:1080 Uncaught (in promise) TypeError: text.split is not a function
    at Function._encode_text (tokenizers.js:1080:1)
    at Function.encode (tokenizers.js:1105:1)
    at tokenizers.js:970:1
    at Array.map (<anonymous>)
    at Function._call (tokenizers.js:970:1)
    at Function.closure [as tokenizer] (utils.js:372:1)
    at Function._call (pipelines.js:214:1)
    at closure (utils.js:372:1)
    at basicStateReducer (react-dom.development.js:16540:1)
    at updateReducer (react-dom.development.js:16664:1)
   ```

**Comment by xenova** [comment]
Hi! This is because the you are using `setState` on a function, which (I believe) calls the function under the hood.

I'm by no means a react expert, but you can get around this by saving 


```tsx
import React, { useState } from 'react';
import { pipeline } from '@xenova/transformers';

/*
Very basic sample react app for transformers.js, showing how to store the model as a promise.

NOTE: This will work for testing, but since the prediction is running in the main thread, the
UI will freeze when loading and running the model. The correct way to handle this is by using
a web or service worker, to offload those intensive operations to a separate thread.
*/
function App() {
  // Stores the model's function as a promise
  const [modelPromise, setModelPromise] = useState<Promise<any> | null>(null);

  // Stores the input text
  const [inputText, setInputText] = useState('');

  // Stores the result/preduction
  const [resultText, setResultText] = useState('');

  const handleInputChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputText(event.target.value);

    let model = await modelPromise;
    if (!model) {
      // The model hasn't been loaded yet. So, we load it here
      let modelPromise = pipeline('text-classification');
      setModelPromise(modelPromise);
      model = await modelPromise;
    }

    setResultText(await model(event.target.value))
  };

  return (
    <div>
      <label htmlFor="input-text">Input Text:</label>
      <input id="input-text" type="text" value={inputText} onChange={handleInputChange} />
      {resultText && (
        <p>{JSON.stringify(resultText)}</p>
      )}
    </div>
  );
}

export default App;
```

---

**Comment by xenova** [comment]
@phamquyhai did the above code solve your problem?

**Comment by DavidGOrtega** [comment]
> @phamquyhai did the above code solve your problem?

Sorry to say that code does not look correct @xenova. Its instantiating a new model and inferring everytime is typing. He needs to use useEffect to create the model once and do the infer after finished typing like you would do in a rest api with setTimeout.

I would be able to help. Im quite proficient in react. But I would need to see what he wants to build.

**Comment by phamquyhai** [comment]
this my hooks

```
// @ts-ignore
import { pipeline } from "@xenova/transformers";
import { useEffect, useState } from "react";

type TASK =
  | "text-classification"
  | "question-answering"
  | "fill-mask"
  | "summarization"
  | "translation"
  | "text2text-generation"
  | "text-generation"
  | "automatic-speech-recognition"
  | "image-to-text"
  | "image-classification"
  | "zero-shot-image-classification"
  | "embeddings"
  | "sentiment-analysis"
  | "ner"
  | "vqa";

export const useTransformers = (task: TASK, model?: string | null) => {
  const [classifier, setClassifier] = useState<any>(null);
  const [status, setStatus] = useState<"pending" | "done" | "error">("pending");
  useEffect(() => {
    (async () => {
      setStatus("pending");
      try {
        const classifier = await pipeline(task, model || null, {
          progress_callback: ({
            progress,
            ...args
          }: {
            progress: number;
            loaded: boolean;
            total: number;
          }) => {
            console.log("progress args : ",args);
          },
        });
        setClassifier(classifier);
        setStatus("done");
      } catch (error) {
        setStatus("error");
      }
    })();
  }, [model, task]);

  return { classifier, status };
};
```

Using this hook
```
const { classifier, status } = useTransformers("translation");
  console.log("status", status);
```

**Comment by xenova** [comment]
Hi @phamquyhai, any updates?

**Comment by DavidGOrtega** [comment]
> tokenizers.js:1080 Uncaught (in promise) TypeError: text.split is not a function

Are you adding the wasm files to your public folder?

**Comment by lucabeetz** [comment]
Hey, thanks a lot for this amazing library!

I am using the above example code to run the whisper-base model inside an electron app. Could you tell me what I would have to change / add to make the model run in a web/service worker?

**Comment by xenova** [comment]
Sure! I would recommend looking at the demo web worker script here:
https://github.com/xenova/transformers.js/blob/main/assets/js/worker.js

There are many ways to go about using the library, but probably the easiest is to use a [singleton](https://en.wikipedia.org/wiki/Singleton_pattern) object for your app. This just ensures that you only create the pipeline once and creation is done lazily (the first time it is called).

**Comment by lucabeetz** [comment]
Thanks for the quick reply! Will try this tomorrow and let you know how it goes

**Comment by xenova** [comment]
Closing since original issue is not a problem with the library, rather it is an issue with how it was used. Also because the original poster has not responded in 3 weeks.

@lucabeetz Feel free to open a new issue if you wish to continue our discussion.
